File: api_nestjs_distributed_order_platform/src/presentation/http/stock.controller.ts
import { Body, Controller, Get, Param, Post, Put } from "@nestjs/common";
import { GetStockUseCase } from "src/application/use-case/stock/get-stock.use-case";
import { ReleaseStockUseCase } from "src/application/use-case/stock/release-stock.use-case";
import { ReserveStockUseCase } from "src/application/use-case/stock/reserve-stock.use-case";
import { UpdateStockUseCase } from "src/application/use-case/stock/update-stock.use-case";
import { toStockResponse } from "./mappers/stock.mapper";
import { UpdateStockDto } from "./dto/stock.dto";
import { StockQuantityDto } from "./dto/stock.dto";
import { ApiOperation, ApiResponse, ApiTags } from "@nestjs/swagger";

@ApiTags('Stock')
@Controller('stock')
export class StockController {
    constructor(
        private readonly getStockUseCase: GetStockUseCase,
        private readonly updateStockUseCase: UpdateStockUseCase,
        private readonly reserveStockUseCase: ReserveStockUseCase,
        private readonly releaseStockUseCase: ReleaseStockUseCase,
    ) {}

    @Get(':productId')
    @ApiOperation({ summary: 'Obtém o estoque de um produto por ID' })
    @ApiResponse({ status: 200, description: 'Estoque obtido com sucesso' })
    @ApiResponse({ status: 400, description: 'Erro ao obter estoque' })
    async getByProductId(@Param('productId') productId: string) {
        const stock = await this.getStockUseCase.execute(productId);
        return toStockResponse(stock);
    }

    @Put(':productId')
    @ApiOperation({ summary: 'Atualiza o estoque de um produto por ID' })
    @ApiResponse({ status: 200, description: 'Estoque atualizado com sucesso' })
    @ApiResponse({ status: 400, description: 'Erro ao atualizar estoque' })
    async update(
        @Param('productId') productId: string,
        @Body() body: UpdateStockDto,
    ) {
        const stock = await this.updateStockUseCase.execute({
            productId,
            availableQuantity: body.availableQuantity,
            reservedQuantity: body.reservedQuantity,
        });
        return toStockResponse(stock);
    }

    @Post(':productId/reserve')
    @ApiOperation({ summary: 'Reserva o estoque de um produto por ID' })
    @ApiResponse({ status: 200, description: 'Estoque reservado com sucesso' })
    @ApiResponse({ status: 400, description: 'Erro ao reservar estoque' })
    async reserve(
        @Param('productId') productId: string,
        @Body() body: StockQuantityDto,
    ) {
        const stock = await this.reserveStockUseCase.execute(
            productId,
            body.quantity,
        );
        return toStockResponse(stock);
    }

    @Post(':productId/release')
    @ApiOperation({ summary: 'Libera o estoque de um produto por ID' })
    @ApiResponse({ status: 200, description: 'Estoque liberado com sucesso' })
    @ApiResponse({ status: 400, description: 'Erro ao liberar estoque' })
    async release(
        @Param('productId') productId: string,
        @Body() body: StockQuantityDto,
    ) {
        const stock = await this.releaseStockUseCase.execute(
            productId,
            body.quantity,
        );
        return toStockResponse(stock);
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/presentation/http/StockController.java
package com.project.order.presentation.http;

import com.project.order.application.usecase.stock.GetStockUseCase;
import com.project.order.application.usecase.stock.ReleaseStockUseCase;
import com.project.order.application.usecase.stock.ReserveStockUseCase;
import com.project.order.application.usecase.stock.UpdateStockUseCase;
import com.project.order.presentation.http.dto.StockDto.StockQuantityRequest;
import com.project.order.presentation.http.dto.StockDto.StockResponse;
import com.project.order.presentation.http.dto.StockDto.UpdateStockRequest;
import com.project.order.presentation.http.mapper.StockMapper;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/stock")
public class StockController {

    private final GetStockUseCase getStockUseCase;
    private final UpdateStockUseCase updateStockUseCase;
    private final ReserveStockUseCase reserveStockUseCase;
    private final ReleaseStockUseCase releaseStockUseCase;

    public StockController(
            GetStockUseCase getStockUseCase,
            UpdateStockUseCase updateStockUseCase,
            ReserveStockUseCase reserveStockUseCase,
            ReleaseStockUseCase releaseStockUseCase
    ) {
        this.getStockUseCase = getStockUseCase;
        this.updateStockUseCase = updateStockUseCase;
        this.reserveStockUseCase = reserveStockUseCase;
        this.releaseStockUseCase = releaseStockUseCase;
    }

    @GetMapping("/{productId}")
    public StockResponse getByProductId(@PathVariable String productId) {
        return StockMapper.toResponse(getStockUseCase.execute(productId));
    }

    @PutMapping("/{productId}")
    public StockResponse update(@PathVariable String productId, @Valid @RequestBody UpdateStockRequest body) {
        return StockMapper.toResponse(updateStockUseCase.execute(productId, body.availableQuantity(), body.reservedQuantity()));
    }

    @PostMapping("/{productId}/reserve")
    public StockResponse reserve(@PathVariable String productId, @Valid @RequestBody StockQuantityRequest body) {
        return StockMapper.toResponse(reserveStockUseCase.execute(productId, body.quantity()));
    }

    @PostMapping("/{productId}/release")
    public StockResponse release(@PathVariable String productId, @Valid @RequestBody StockQuantityRequest body) {
        return StockMapper.toResponse(releaseStockUseCase.execute(productId, body.quantity()));
    }
}
